### ISSUE TITLE
weird hash type issue

### ISSUE THREAD

**Issue by zseri** [NixOS/nix#6377]
**Describe the bug**

I just rebuilt a nixos system, decided to change a small thing and Nix immediately threw an error at me completely unrelated to any change I made...

```
error: hash 'sha512-PcXTctrO0aL5C7Ci1J2Z5fa5WqKONhOK2q0FnSbT5+iP9WWSCljyQiHE8C4LYMMHii48y6AJVRkjVIukI3+rUQ==' should have type 'sha256'

       … while evaluating the attribute 'buildCommand' of the derivation 'texlive-texdoc-3.3.1'

       at /nix/store/9wrqzbgsxq072pcc8m64lavzv03m64js-source/pkgs/stdenv/generic/make-derivation.nix:205:7:

          204|     // (lib.optionalAttrs (attrs ? name || (attrs ? pname && attrs ? version)) {
          205|       name =
             |       ^
          206|         let

       … while evaluating the attribute 'outPath'

       at /nix/store/9wrqzbgsxq072pcc8m64lavzv03m64js-source/lib/customisation.nix:164:7:

          163|       drvPath = assert condition; drv.drvPath;
          164|       outPath = assert condition; drv.outPath;
             |       ^
          165|     };

       … while evaluating anonymous lambda

       at /nix/store/9wrqzbgsxq072pcc8m64lavzv03m64js-source/pkgs/tools/typesetting/tex/texlive/combine.nix:39:11:

           38|   mkUniqueOutPaths = pkgs: lib.unique
           39|     (map (p: p.outPath) (builtins.filter lib.isDerivation pkgs));
             |           ^
           40|

       … from call site

       … while evaluating anonymous lambda

       at /nix/store/9wrqzbgsxq072pcc8m64lavzv03m64js-source/lib/lists.nix:646:25:

          645|    */
          646|   unique = foldl' (acc: e: if elem e acc then acc else acc ++ [ e ]) [];
             |                         ^
          647|

       … from call site
[..etc..]
```

This might be a `nixpkgs` bug, but I'm not sure, especially because the error is not particularly helpful because it doesn't even point to the introduction point of the hash, that is, where exactly it comes from...

**Steps To Reproduce**

Try to build https://github.com/zseri/nixos-configs/commit/2c59a5f6a5335d106c7049745e5fa4adcebb1b7b with the specified Nix version and flake attribute `#nixosConfigurations.orca.config.system.build.toplevel`. 

**Expected behavior**
The error message should point to the string definition location. And evaluation caching seems to affect when the error appears.

**`nix-env --version` output**
broken:
`nix-env (Nix) 2.8.0pre20220401_a24d077`
works as expected:
`nix-env (Nix) 2.8.0pre20220325_1844172`

**Additional context**
`experimental-features = nix-command ca-derivations flakes`

**Comment by zseri** [comment]
It appeared in nixpkgs here:
```
pkgs/tools/typesetting/tex/texlive/default.nix
139:    hash = "sha512-PcXTctrO0aL5C7Ci1J2Z5fa5WqKONhOK2q0FnSbT5+iP9WWSCljyQiHE8C4LYMMHii48y6AJVRkjVIukI3+rUQ==";
```

Also, the issue is not present with a newer nixpkgs rev, so it certainly got fixed...
But why did the old version of Nix not display the error?

**Comment by edolstra** [comment]
This is fixed in #6347.
